### ISSUE TITLE
Use different backbones

### ISSUE THREAD

**Issue by EckoTan0804** [yangsenius/TransPose#21]
Hello, 

I want to replace ResNet in TransPose-R with different backbones. How should I do it correctly? Should I import the backbone and replace the following code with the forward pass of the new backbone?

https://github.com/yangsenius/TransPose/blob/dab9007b6f61c9c8dce04d61669a04922bbcd148/lib/models/transpose_r.py#L403-L410

Thanks in advance!

**Comment by yangsenius** [comment]
Yes, you are right. You also need to ensure that the downsampling ratio of position embedding and upsampling layers coordinate with the backbone. [position embedding](https://github.com/yangsenius/TransPose/blob/dab9007b6f61c9c8dce04d61669a04922bbcd148/lib/models/transpose_r.py#L303), [upsampling layers](https://github.com/yangsenius/TransPose/blob/dab9007b6f61c9c8dce04d61669a04922bbcd148/lib/models/transpose_r.py#L303)

**Comment by EckoTan0804** [comment]
Thanks for your reply!

Currently I am trying to apply a new backbone for TransPose-R. The new backbone adopts 32x downsampling ratio and outputs feature maps of size HxW=8x4 for 256x192 input. However, the [position embedding](https://github.com/yangsenius/TransPose/blob/dab9007b6f61c9c8dce04d61669a04922bbcd148/lib/models/transpose_r.py#L303) of TransPose-R has an 8x downsampling ratio. Should I directly modify the positional embedding as

```python
self.pe_h = h // 32
self.pe_w = w // 32
```
, or should I apply a 4x upsampling for the feature maps?

Btw [position embedding](https://github.com/yangsenius/TransPose/blob/dab9007b6f61c9c8dce04d61669a04922bbcd148/lib/models/transpose_r.py#L303), [upsampling layers](https://github.com/yangsenius/TransPose/blob/dab9007b6f61c9c8dce04d61669a04922bbcd148/lib/models/transpose_r.py#L303) you mentioned above seem to refer to the same line of code.

**Comment by yangsenius** [comment]
The position embedding part is right. You should modify the config of deconvolution layers [upsampling layers](https://github.com/yangsenius/TransPose/blob/dab9007b6f61c9c8dce04d61669a04922bbcd148/lib/models/transpose_r.py#L282) by changing the [YAML: NUM_DECONV_LAYERS;NUM_DECONV_FILTERS;NUM_DECONV_KERNELS](https://github.com/yangsenius/TransPose/blob/dab9007b6f61c9c8dce04d61669a04922bbcd148/experiments/coco/transpose_r/TP_R_256x192_d256_h1024_enc4_mh8.yaml#L48). 8x unsampling via 3 deconvolutioal layers are needed to recover the 1/32 input size to 1/4 size.

**Comment by EckoTan0804** [comment]
> The position embedding part is right. You should modify the config of deconvolution layers [upsampling layers](https://github.com/yangsenius/TransPose/blob/dab9007b6f61c9c8dce04d61669a04922bbcd148/lib/models/transpose_r.py#L282) by changing the [YAML: NUM_DECONV_LAYERS;NUM_DECONV_FILTERS;NUM_DECONV_KERNELS](https://github.com/yangsenius/TransPose/blob/dab9007b6f61c9c8dce04d61669a04922bbcd148/experiments/coco/transpose_r/TP_R_256x192_d256_h1024_enc4_mh8.yaml#L48). 8x unsampling via 3 deconvolutioal layers are needed to recover the 1/32 input size to 1/4 size.

Thanks for your suggestion. 
Just to confirm, to apply my new backbone (which adopts 32x downsampling ratio) for TransPose-R, I need to 

1. Adjust the dimension of positional embedding:
```python
self.pe_h = h // 32
self.pe_w = w // 32
```

2. Modify the config of DECONV layers as
```yaml
NUM_DECONV_LAYERS: 3
NUM_DECONV_FILTERS:
- 256
- 256
- 256
NUM_DECONV_KERNELS:
- 4
- 4
- 4
```


3. Replace the forward pass of ResNet backbone with my new backbone


If there's mistake, please correct me.
